src/model/core/fuel.jl

```diff
@@ -0,0 +1,150 @@
+        if y in THERM_COMMIT
+            (dfGen[y,:Cap_Size] * EP[:vSTART][y, t] * 
+                dfGen[y,:Start_Fuel_MMBTU_per_MW])
+        else
+            0
+        end)
+
+    # fuel_cost is in $/MMBTU (M$/billion BTU if scaled)
+    # vFuel and eStartFuel is MMBTU (or billion BTU if scaled)
+    # eCFuel_start or eCFuel_out is $ or Million$
+    # Start up fuel cost
+    @expression(EP, eCFuelStart[y = 1:G, t = 1:T], 
+        (inputs["fuel_costs"][dfGen[y,:Fuel]][t] * EP[:eStartFuel][y, t]))
+    # plant level start-up fuel cost for output
+    @expression(EP, ePlantCFuelStart[y = 1:G], 
+        sum(inputs["omega"][t] * EP[:eCFuelStart][y, t] for t in 1:T))
+    # zonal level total fuel cost for output
+    @expression(EP, eZonalCFuelStart[z = 1:Z], 
+        sum(EP[:ePlantCFuelStart][y] for y in dfGen[dfGen[!, :Zone].==z, :R_ID]))
+
+    # Fuel cost for power generation
+    @expression(EP, eCFuelOut[y = 1:G, t = 1:T], 
+        (inputs["fuel_costs"][dfGen[y,:Fuel]][t] * EP[:vFuel][y, t]))
+    # plant level start-up fuel cost for output
+    @expression(EP, ePlantCFuelOut[y = 1:G], 
+        sum(inputs["omega"][t] * EP[:eCFuelOut][y, t] for t in 1:T))
+    # zonal level total fuel cost for output
+    @expression(EP, eZonalCFuelOut[z = 1:Z], 
+        sum(EP[:ePlantCFuelOut][y] for y in dfGen[dfGen[!, :Zone].==z, :R_ID]))
+
+
+    # system level total fuel cost for output
+    @expression(EP, eTotalCFuelOut, sum(eZonalCFuelOut[z] for z in 1:Z))
+    @expression(EP, eTotalCFuelStart, sum(eZonalCFuelStart[z] for z in 1:Z))
+
+
+    add_to_expression!(EP[:eObj], EP[:eTotalCFuelOut] + EP[:eTotalCFuelStart])
+
+    #fuel consumption (MMBTU or Billion BTU)
+    @expression(EP, eFuelConsumption[f in 1:NUM_FUEL, t in 1:T],
+        sum(EP[:vFuel][y, t] + EP[:eStartFuel][y,t]
+            for y in resources_with_fuel(dfGen, fuels[f])))
+                
+    @expression(EP, eFuelConsumptionYear[f in 1:NUM_FUEL],
+        sum(inputs["omega"][t] * EP[:eFuelConsumption][f, t] for t in 1:T))
+
+    
+    ### Constraint ###
+    ### only apply constraint to generators with fuel type other than None
+    @constraint(EP, FuelCalculation[y in setdiff(HAS_FUEL, THERM_COMMIT), t = 1:T],
+        EP[:vFuel][y, t] - EP[:vP][y, t] * dfGen[y, :Heat_Rate_MMBTU_per_MWh] == 0)
+
+    if !isempty(THERM_COMMIT)
+        if setup["PiecewiseFuelUsage"] == 1
+            # Only apply piecewise fuel consumption to thermal generators with PWFU_NUM_SEGMENTS > 0
+            THERM_COMMIT_PWFU = inputs["THERM_COMMIT_PWFU"]
+            for segment in 1:inputs["PWFU_MAX_NUM_SEGMENTS"]
+                @constraint(EP, [y in THERM_COMMIT_PWFU, t = 1:T],
+                EP[:vFuel][y, t] >= (EP[:vP][y, t] *  dfGen[y, inputs["slope_cols"]][segment] + 
+                    EP[:vCOMMIT][y, t] * dfGen[y, inputs["intercept_cols"]][segment]))
+            end
+            @constraint(EP, [y in setdiff(THERM_COMMIT,THERM_COMMIT_PWFU), t = 1:T],
```

Comment:
Please add a constraint name

Reply:
I don't know how to do that in a for loop, it gives me an error message like constraint names are already in the model. Any idea @cfe316 ?